# Exploratory Data Analysis (EDA) for Givemore

Import Libraries

In [1]:
import pandas as pd

Load Data

In [5]:
ratings = pd.read_csv("/home/bukunmi/code/givemore/data/ratings.csv")
ratings.head()

,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


In [11]:
movies = pd.read_csv("/home/bukunmi/code/givemore/data/movies.csv")
movies.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


### **1.** What's the exact density of the user-movie matrix, and what does it imply for the minimum number of co-ratings before a similarity is meaningful?

  $$\text{density} = \frac{\text{ratings}}{\text{users} \times \text{movies}}$$

In [18]:
num_ratings = ratings["rating"].count()
print(f"Number of ratings: {num_ratings}")
num_users = ratings["userId"].nunique()
print(f"Number of users: {num_users}")
num_movies = movies["movieId"].nunique()
print(f"Number of movies: {num_movies}")

Number of ratings: 100836
Number of users: 610
Number of movies: 9742


In [17]:
density = num_ratings / (num_users * num_movies)
print(f"Density: {(density*100):.2f}%")

Density: 1.70%


At **_1.70%_** density the matrix is **_98.3%_** empty. The average user has rated 165 of 9,742 movies, and typical movie pairs share 0 raters. Behavioural (collaborative) similarity is therefore only reliable for the popular head; the long tail needs a content-based fallback and a minimum co-rating threshold to avoid spurious similarities.